auth/proofOfPossession.go

```diff
@@ -0,0 +1,83 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Cloud Security Client Go contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+package auth
+
+// parseAndValidateCertificate checks proof of possession in addition to audience validation
+// to make sure that it was called by a trust-worthy consumer.
+// Trust between application and applications/services is established with certificates in principle.
+// Proof of possession uses certificates as proof token and therefore, x.509 based mTLS communication is demanded.
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+)
+
+var ErrNoClientCert = errors.New("there is no x509 client certificate provided")
+var ErrNoToken = errors.New("there is no token provided")
+
+func parseAndValidateCertificate(clientCertificate *x509.Certificate, token Token) error {
+	if clientCertificate == nil {
+		return ErrNoClientCert
+	}
+	if token == nil {
+		return ErrNoToken
+	}
+	return ValidateX5tThumbprint(clientCertificate, token)
+}
+
+// In order to check whether the token was issued for the sender,
+// the cnf token claim with the confirmation method “x5t#S256” needs to be compared with the thumbprint of the
+// provided X509 client certificate.
+// See also RFC 8705
+func ValidateX5tThumbprint(clientCertificate *x509.Certificate, token Token) error {
+	if clientCertificate == nil {
+		return ErrNoClientCert
+	}
+	if token == nil {
+		return ErrNoToken
+	}
+
+	cnfThumbprint := token.getCnfClaimMember(claimCnfMemberX5t)
+	if cnfThumbprint == "" {
+		return fmt.Errorf("token provides no cnf member for thumbprint confirmation")
+	}
+
+	certThumbprintBytes := sha256.Sum256(clientCertificate.Raw)
+	certThumbprint := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(certThumbprintBytes[:])
+
+	if cnfThumbprint != certThumbprint {
+		return fmt.Errorf("token thumbprint confirmation failed")
+	}
+	return nil
+}
+
+// Parses the X509 client certificate which is provided via the "x-forwarded-client-cert".
+// It supports DER encoded and PEM encoded certificates.
+// Returns nil, if certString is empty string.
+func parseCertString(certString string) (*x509.Certificate, error) {
```

Comment:
in java I renamed it to parseCertificate(...) to have consistent usage of `certificate` across all classes and fields

Reply:
done